src/main.rs

```diff
@@ -36,41 +54,44 @@ fn sector_count(size: u64, sector_size: u64) -> u64 {
     size / sector_size
 }
 
-fn main() -> io::Result<()> {
+fn main() -> Result<()> {
     let args = Args::parse();
 
     let root = Path::new(&args.filename).parent().unwrap();
-    let cue_sheet = std::fs::read_to_string(&args.filename)?;
+    let cue_sheet = std::fs::read_to_string(&args.filename).into_diagnostic()?;
 
-    let cd = CD::parse(cue_sheet)?;
+    let cd = CD::parse(cue_sheet).into_diagnostic()?;
 
     let tracks = cd.tracks();
     if has_multiple_files(tracks) {
-        println!("This tool currently only supports single-file BIN/CUE images.");
-        exit(1);
+        return Err(MultipleFilesError { cue: args.filename })?;
     }
     let fname = cd.tracks().first().unwrap().get_filename();
     let file = root.join(fname);
     if !file.is_file() {
-        println!("Cuesheet file {} does not exist", file.to_string_lossy());
-        exit(1);
+        return Err(MissingFileError {
+            cue: args.filename,
+            missing_file: file.to_string_lossy().to_string(),
```

Comment:
Could also store the proper PathBuf here and print `path.display()` in the `#[error(...)]` (I pervasively push for camino::Utf8Path so I don't have to do this dance...)

Reply:
Good call, updated.

I *usually* expect ASCII paths here, but given what an old file format it is, it's not guaranteed I'm going to run into UTF-8 paths... 🙈